File: desktop/src-tauri/src/dlc_asset_protocol.rs
use serde::{Deserialize, Serialize};
use std::borrow::Cow;
use std::path::{Path, PathBuf};
use std::sync::{Arc, RwLock};
use tauri::http::{header, Request, Response, StatusCode};

pub const MAX_DLC_ASSET_BYTES: usize = 20 * 1024 * 1024; // 20 MiB
const PROTOCOL_SCHEME: &str = "dlc-asset";

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActiveDlcProjectionItem {
    pub dlc_id: String,
    pub package_version: String,
    pub package_digest: String,
    pub frontend_entrypoint: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RuntimeDlcActivationProjection {
    pub snapshot_id: String,
    pub active_dlcs: Vec<ActiveDlcProjectionItem>,
}

#[derive(Debug, Default, Clone)]
pub struct DlcAssetHostState {
    inner: Arc<RwLock<Option<RuntimeDlcActivationProjection>>>,
}

impl DlcAssetHostState {
    pub fn new() -> Self {
        Self {
            inner: Arc::new(RwLock::new(None)),
        }
    }

    pub fn update_projection(&self, projection: RuntimeDlcActivationProjection) {
        if let Ok(mut lock) = self.inner.write() {
            *lock = Some(projection);
        }
    }

    pub fn clear(&self) {
        if let Ok(mut lock) = self.inner.write() {
            *lock = None;
        }
    }

    pub fn is_digest_active(&self, digest: &str) -> bool {
        let lock = match self.inner.read() {
            Ok(guard) => guard,
            Err(_) => return false,
        };
        let Some(projection) = lock.as_ref() else {
            return false;
        };
        let normalized = normalize_digest(digest);
        projection.active_dlcs.iter().any(|d| {
            normalize_digest(&d.package_digest).eq_ignore_ascii_case(&normalized)
        })
    }

    #[allow(dead_code)]
    pub fn current_projection(&self) -> Option<RuntimeDlcActivationProjection> {
        self.inner.read().ok().and_then(|guard| guard.clone())
    }
}

pub fn normalize_digest(digest: &str) -> String {
    digest
        .strip_prefix("sha256:")
        .or_else(|| digest.strip_prefix("sha256-"))
        .unwrap_or(digest)
        .to_ascii_lowercase()
}

#[derive(Debug, PartialEq, Eq)]
pub struct ParsedDlcAssetRequest {
    pub package_digest: String,
    pub subpath: String,
}

pub fn parse_dlc_asset_url(raw_url: &str) -> Result<ParsedDlcAssetRequest, String> {
    let url = tauri::Url::parse(raw_url).map_err(|e| format!("Invalid URL: {e}"))?;
    if url.scheme() != PROTOCOL_SCHEME {
        return Err(format!("Invalid scheme '{}': expected '{PROTOCOL_SCHEME}'", url.scheme()));
    }
    if let Some(host) = url.host_str() {
        if !host.is_empty() && !host.eq_ignore_ascii_case("localhost") && host != "127.0.0.1" {
            return Err("Invalid host: only localhost is permitted".to_string());
        }
    }
    let mut segments = url
        .path_segments()
        .ok_or_else(|| "Missing path segments in dlc-asset URL".to_string())?;

    let raw_digest = segments
        .next()
        .ok_or_else(|| "Missing package digest in dlc-asset URL".to_string())?;
    let normalized_digest = normalize_digest(raw_digest);
    if normalized_digest.len() != 64 || !normalized_digest.chars().all(|c| c.is_ascii_hexdigit()) {
        return Err("Invalid package digest format: expected 64 hex characters".to_string());
    }

    let mut remaining: Vec<&str> = Vec::new();
    for segment in segments {
        if segment.is_empty() || segment == "." {
            continue;
        }
        if segment == ".."
            || segment.contains('\\')
            || segment.contains('\0')
            || segment.contains('/')
        {
            return Err("Path traversal segment rejected".to_string());
        }
        remaining.push(segment);
    }

    if remaining.is_empty() {
        return Err("Missing asset subpath".to_string());
    }

    // Strip leading "frontend" if present in URL path (since base dir is already /frontend/)
    let subpath = if remaining.first().copied() == Some("frontend") {
        remaining[1..].join("/")
    } else {
        remaining.join("/")
    };

    if subpath.is_empty() {
        return Err("Missing frontend asset subpath".to_string());
    }

    Ok(ParsedDlcAssetRequest {
        package_digest: normalized_digest,
        subpath,
    })
}

pub fn resolve_dlc_packages_root() -> PathBuf {
    if let Ok(override_dir) = std::env::var("DBFOX_RUNTIME_DIR") {
        if !override_dir.trim().is_empty() {
            return PathBuf::from(override_dir).join("dlcs").join("packages");
        }
    }

    #[cfg(target_os = "windows")]
    {
        if let Ok(appdata) = std::env::var("APPDATA") {
            return PathBuf::from(appdata)
                .join("DBFox")
                .join("dlcs")
                .join("packages");
        }
    }

    #[cfg(target_os = "macos")]
    {
        if let Ok(home) = std::env::var("HOME") {
            return PathBuf::from(home)
                .join("Library")
                .join("Application Support")
                .join("DBFox")
                .join("dlcs")
                .join("packages");
        }
    }

    #[cfg(not(any(target_os = "windows", target_os = "macos")))]
    {
        if let Ok(xdg) = std::env::var("XDG_DATA_HOME") {
            return PathBuf::from(xdg).join("dbfox").join("dlcs").join("packages");
        }
        if let Ok(home) = std::env::var("HOME") {
            return PathBuf::from(home)
                .join(".local")
                .join("share")
                .join("dbfox")
                .join("dlcs")
                .join("packages");
        }
    }

    PathBuf::from("dlcs").join("packages")
}

pub fn mime_for_path(path: &Path) -> &'static str {
    match path.extension().and_then(|s| s.to_str()).unwrap_or("").to_ascii_lowercase().as_str() {
        "js" | "mjs" => "text/javascript; charset=utf-8",
        "css" => "text/css; charset=utf-8",
        "json" => "application/json; charset=utf-8",
        "svg" => "image/svg+xml",
        "png" => "image/png",
        "jpg" | "jpeg" => "image/jpeg",
        "webp" => "image/webp",
        "gif" => "image/gif",
        "ico" => "image/x-icon",
        "woff" => "font/woff",
        "woff2" => "font/woff2",
        "ttf" => "font/ttf",
        "otf" => "font/otf",
        "html" => "text/html; charset=utf-8",
        _ => "application/octet-stream",
    }
}

fn make_text_response(status: StatusCode, body: &'static str) -> Response<Cow<'static, [u8]>> {
    Response::builder()
        .status(status)
        .header(header::CONTENT_TYPE, "text/plain; charset=utf-8")
        .header(header::ACCESS_CONTROL_ALLOW_ORIGIN, "*")
        .header(header::CACHE_CONTROL, "no-cache")
        .header("X-Content-Type-Options", "nosniff")
        .body(Cow::Borrowed(body.as_bytes()))
        .unwrap_or_else(|_| Response::new(Cow::Borrowed(&[] as &'static [u8])))
}

pub fn handle_dlc_asset_request(
    state: &DlcAssetHostState,
    request: Request<Vec<u8>>,
) -> Response<Cow<'static, [u8]>> {
    let uri_str = request.uri().to_string();

    // Handle CORS preflight
    if request.method() == "OPTIONS" {
        return Response::builder()
            .status(StatusCode::NO_CONTENT)
            .header(header::ACCESS_CONTROL_ALLOW_ORIGIN, "*")
            .header(header::ACCESS_CONTROL_ALLOW_METHODS, "GET, HEAD, OPTIONS")
            .header(header::ACCESS_CONTROL_ALLOW_HEADERS, "*")
            .body(Cow::Borrowed(&[] as &'static [u8]))
            .unwrap_or_else(|_| Response::new(Cow::Borrowed(&[] as &'static [u8])));
    }

    if request.method() != "GET" && request.method() != "HEAD" {
        return make_text_response(StatusCode::METHOD_NOT_ALLOWED, "Method not allowed");
    }

    let parsed = match parse_dlc_asset_url(&uri_str) {
        Ok(parsed) => parsed,
        Err(_) => {
            return make_text_response(StatusCode::BAD_REQUEST, "Invalid DLC asset URL");
        }
    };

    // 1. Verify active projection allowlist (Snapshot is active truth)
    if !state.is_digest_active(&parsed.package_digest) {
        return make_text_response(
            StatusCode::FORBIDDEN,
            "DLC package is not active in the current runtime snapshot",
        );
    }

    // 2. Resolve package directory and enforce containment
    let packages_root = resolve_dlc_packages_root();
    let base_dir = packages_root
        .join(format!("sha256-{}", parsed.package_digest))
        .join("frontend");

    let canonical_base = match base_dir.canonicalize() {
        Ok(path) => path,
        Err(_) => {
            return make_text_response(StatusCode::NOT_FOUND, "DLC package frontend assets not found");
        }
    };

    let target_path = base_dir.join(&parsed.subpath);
    let canonical_target = match target_path.canonicalize() {
        Ok(path) => path,
        Err(_) => {
            return make_text_response(StatusCode::NOT_FOUND, "Asset file not found");
        }
    };

    if !canonical_target.starts_with(&canonical_base) {
        return make_text_response(StatusCode::FORBIDDEN, "Path traversal forbidden");
    }

    if !canonical_target.is_file() {
        return make_text_response(StatusCode::NOT_FOUND, "Requested path is not a file");
    }

    // 3. Check bounded size limit
    let metadata = match std::fs::metadata(&canonical_target) {
        Ok(meta) => meta,
        Err(_) => {
            return make_text_response(
                StatusCode::INTERNAL_SERVER_ERROR,
                "Failed to read asset metadata",
            );
        }
    };

    if metadata.len() > MAX_DLC_ASSET_BYTES as u64 {
        return make_text_response(
            StatusCode::PAYLOAD_TOO_LARGE,
            "Asset file exceeds 20 MiB maximum size",
        );
    }

    // 4. Read bytes
    let bytes = match std::fs::read(&canonical_target) {
        Ok(data) => data,
        Err(_) => {
            return make_text_response(
                StatusCode::INTERNAL_SERVER_ERROR,
                "Failed to read asset file",
            );
        }
    };

    let mime_type = mime_for_path(&canonical_target);

    Response::builder()
        .status(StatusCode::OK)
        .header(header::CONTENT_TYPE, mime_type)
        .header(header::ACCESS_CONTROL_ALLOW_ORIGIN, "*")
        .header(header::CACHE_CONTROL, "public, max-age=31536000, immutable")
        .header("X-Content-Type-Options", "nosniff")
        .body(Cow::Owned(bytes))
        .unwrap_or_else(|_| Response::new(Cow::Borrowed(&[] as &'static [u8])))
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn parses_valid_dlc_asset_urls() {
        let digest = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";
        let url = format!("dlc-asset://localhost/{digest}/frontend/index.js");
        let parsed = parse_dlc_asset_url(&url).unwrap();
        assert_eq!(parsed.package_digest, digest);
        assert_eq!(parsed.subpath, "index.js");

        let url_no_frontend_prefix = format!("dlc-asset://localhost/{digest}/assets/icon.png");
        let parsed2 = parse_dlc_asset_url(&url_no_frontend_prefix).unwrap();
        assert_eq!(parsed2.package_digest, digest);
        assert_eq!(parsed2.subpath, "assets/icon.png");
    }

    #[test]
    fn rejects_invalid_schemes_and_hosts() {
        let digest = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";
        assert!(parse_dlc_asset_url(&format!("http://localhost/{digest}/index.js")).is_err());
        assert!(parse_dlc_asset_url(&format!("dlc-asset://evil.com/{digest}/index.js")).is_err());
        assert!(parse_dlc_asset_url(&format!("dlc-asset://localhost/short_digest/index.js")).is_err());
    }

    #[test]
    fn rejects_path_traversal_segments() {
        let digest = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";
        assert!(parse_dlc_asset_url(&format!("dlc-asset://localhost/{digest}/../manifest.json")).is_err());
        assert!(parse_dlc_asset_url(&format!("dlc-asset://localhost/{digest}/frontend/../../backend/entry.py")).is_err());
    }

    #[test]
    fn host_state_enforces_active_projection_allowlist() {
        let state = DlcAssetHostState::new();
        let digest = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";

        // Uninitialized state rejects
        assert!(!state.is_digest_active(digest));

        // Update with active projection
        state.update_projection(RuntimeDlcActivationProjection {
            snapshot_id: "snap_test".to_string(),
            active_dlcs: vec![ActiveDlcProjectionItem {
                dlc_id: "acme.test".to_string(),
                package_version: "1.0.0".to_string(),
                package_digest: format!("sha256:{digest}"),
                frontend_entrypoint: Some("frontend/index.js".to_string()),
            }],
        });

        assert!(state.is_digest_active(digest));
        assert!(state.is_digest_active(&format!("sha256:{digest}")));
        assert!(!state.is_digest_active("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"));

        // Clear invalidates
        state.clear();
        assert!(!state.is_digest_active(digest));
    }

    #[test]
    fn resolves_correct_mime_types() {
        assert_eq!(mime_for_path(Path::new("index.js")), "text/javascript; charset=utf-8");
        assert_eq!(mime_for_path(Path::new("index.mjs")), "text/javascript; charset=utf-8");
        assert_eq!(mime_for_path(Path::new("style.css")), "text/css; charset=utf-8");
        assert_eq!(mime_for_path(Path::new("icon.svg")), "image/svg+xml");
        assert_eq!(mime_for_path(Path::new("icon.png")), "image/png");
        assert_eq!(mime_for_path(Path::new("data.json")), "application/json; charset=utf-8");
        assert_eq!(mime_for_path(Path::new("font.woff2")), "font/woff2");
        assert_eq!(mime_for_path(Path::new("unknown.bin")), "application/octet-stream");
    }

    #[test]
    fn serves_asset_when_active_and_contained() {
        let temp = tempdir().unwrap();
        let digest = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";
        let pkg_frontend = temp
            .path()
            .join("dlcs")
            .join("packages")
            .join(format!("sha256-{digest}"))
            .join("frontend");
        std::fs::create_dir_all(&pkg_frontend).unwrap();
        let js_file = pkg_frontend.join("index.js");
        std::fs::write(&js_file, b"console.log('hello from dlc');").unwrap();

        std::env::set_var("DBFOX_RUNTIME_DIR", temp.path().to_str().unwrap());

        let state = DlcAssetHostState::new();
        state.update_projection(RuntimeDlcActivationProjection {
            snapshot_id: "snap_1".to_string(),
            active_dlcs: vec![ActiveDlcProjectionItem {
                dlc_id: "test.dlc".to_string(),
                package_version: "1.0.0".to_string(),
                package_digest: digest.to_string(),
                frontend_entrypoint: Some("frontend/index.js".to_string()),
            }],
        });

        // Request active asset
        let req = Request::builder()
            .method("GET")
            .uri(format!("dlc-asset://localhost/{digest}/frontend/index.js"))
            .body(Vec::new())
            .unwrap();

        let resp = handle_dlc_asset_request(&state, req);
        assert_eq!(resp.status(), StatusCode::OK);
        assert_eq!(
            resp.headers().get(header::CONTENT_TYPE).unwrap(),
            "text/javascript; charset=utf-8"
        );
        assert_eq!(resp.body().as_ref(), b"console.log('hello from dlc');");

        // Request non-active asset -> 403 Forbidden
        let req_inactive = Request::builder()
            .method("GET")
            .uri("dlc-asset://localhost/ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff/frontend/index.js")
            .body(Vec::new())
            .unwrap();

        let resp_inactive = handle_dlc_asset_request(&state, req_inactive);
        assert_eq!(resp_inactive.status(), StatusCode::FORBIDDEN);
    }
}

File: desktop/main/dlcAssetProtocol.ts
import { readFile, realpath, stat } from "node:fs/promises";
import { homedir } from "node:os";
import { isAbsolute, join, relative } from "node:path";
import type { Protocol } from "electron";

import type { EngineConfig } from "../shared/desktopContract";
import { readBoundedBody } from "./nodeEngineHost";

const MAX_DLC_ASSET_BYTES = 20 * 1024 * 1024;
const MAX_ACTIVATION_RESPONSE_BYTES = 64 * 1024;

export interface ActiveDlcProjectionItem {
  dlc_id: string;
  package_version: string;
  package_digest: string;
  frontend_entrypoint: string | null;
}

export interface RuntimeDlcActivationProjection {
  snapshot_id: string;
  active_dlcs: ActiveDlcProjectionItem[];
}

export class DlcAssetAuthority {
  #projection: RuntimeDlcActivationProjection | null = null;
  #epoch = 0;

  clear(): void {
    this.#epoch += 1;
    this.#projection = null;
  }

  isActive(packageDigest: string): boolean {
    const digest = normalizeDigest(packageDigest);
    return this.#projection?.active_dlcs.some((item) => item.frontend_entrypoint?.trim()
      && normalizeDigest(item.package_digest) === digest) ?? false;
  }

  updateProjection(projection: RuntimeDlcActivationProjection): void {
    this.#epoch += 1;
    this.#projection = null;
    this.#projection = parseProjection(projection);
  }

  async synchronize(
    config: EngineConfig,
    rendererOrigin: string,
  ): Promise<RuntimeDlcActivationProjection> {
    const epoch = ++this.#epoch;
    this.#projection = null;
    const response = await fetch(`http://127.0.0.1:${config.port}/api/v1/dlcs/activation`, {
      headers: { Origin: rendererOrigin, "X-Local-Token": config.token },
      signal: AbortSignal.timeout(3_000),
    });
    const bytes = await readBoundedBody(response, MAX_ACTIVATION_RESPONSE_BYTES);
    if (!response.ok) throw new Error(`DLC activation projection returned HTTP ${response.status}`);
    const projection = parseProjection(JSON.parse(new TextDecoder().decode(bytes)));
    if (epoch === this.#epoch) this.#projection = projection;
    return projection;
  }
}

export function registerDlcAssetProtocol(protocol: Protocol, authority: DlcAssetAuthority): void {
  protocol.handle("dlc-asset", (request) => handleDlcAssetRequest(authority, request));
}

export async function handleDlcAssetRequest(
  authority: DlcAssetAuthority,
  request: Request,
  packagesRoot = resolveDlcPackagesRoot(),
): Promise<Response> {
  if (request.method === "OPTIONS") {
    return new Response(null, { status: 204, headers: corsHeaders("no-cache") });
  }
  if (request.method !== "GET" && request.method !== "HEAD") return textResponse(405, "Method not allowed");
  let parsed: { packageDigest: string; subpath: string };
  try {
    parsed = parseDlcAssetUrl(request.url);
  } catch {
    return textResponse(400, "Invalid DLC asset URL");
  }
  if (!authority.isActive(parsed.packageDigest)) {
    return textResponse(403, "DLC package is not active in the current runtime snapshot");
  }
  if (packagesRoot === null) return textResponse(404, "DLC runtime directory is unavailable");
  try {
    const base = await realpath(join(packagesRoot, `sha256-${parsed.packageDigest}`, "frontend"));
    const target = await realpath(join(base, parsed.subpath));
    const child = relative(base, target);
    if (child === "" || child === ".." || child.startsWith("../") || child.startsWith("..\\") || isAbsolute(child)) {
      return textResponse(403, "Path traversal forbidden");
    }
    const metadata = await stat(target);
    if (!metadata.isFile()) return textResponse(404, "Requested path is not a file");
    if (metadata.size > MAX_DLC_ASSET_BYTES) return textResponse(413, "Asset file exceeds 20 MiB maximum size");
    const headers = corsHeaders("public, max-age=31536000, immutable");
    headers.set("Content-Type", mimeForPath(target));
    if (request.method === "HEAD") return new Response(null, { status: 200, headers });
    const bytes = await readFile(target);
    return new Response(bytes, { status: 200, headers });
  } catch (error) {
    if (isNodeError(error, "ENOENT")) return textResponse(404, "Asset file not found");
    return textResponse(500, "Failed to read asset file");
  }
}

export function parseDlcAssetUrl(rawUrl: string): { packageDigest: string; subpath: string } {
  const url = new URL(rawUrl);
  if (url.protocol !== "dlc-asset:") throw new Error("Invalid scheme");
  if (url.hostname && url.hostname !== "127.0.0.1" && url.hostname.toLowerCase() !== "localhost") {
    throw new Error("Invalid host");
  }
  const segments = url.pathname.split("/").filter(Boolean).map((segment) => decodeURIComponent(segment));
  const packageDigest = normalizeDigest(segments.shift() ?? "");
  if (!/^[0-9a-f]{64}$/.test(packageDigest)) throw new Error("Invalid package digest");
  if (segments[0] === "frontend") segments.shift();
  if (segments.length === 0 || segments.some((segment) => segment === "." || segment === ".."
    || segment.includes("/") || segment.includes("\\") || segment.includes("\0"))) {
    throw new Error("Invalid asset subpath");
  }
  return { packageDigest, subpath: segments.join("/") };
}

export function resolveDlcPackagesRoot(env = process.env): string | null {
  if (env.DBFOX_RUNTIME_DIR?.trim()) return join(env.DBFOX_RUNTIME_DIR, "dlcs", "packages");
  if (process.platform === "win32" && env.APPDATA?.trim()) return join(env.APPDATA, "DBFox", "dlcs", "packages");
  if (process.platform === "darwin") return join(homedir(), "Library", "Application Support", "DBFox", "dlcs", "packages");
  if (process.platform !== "win32") {
    const root = env.XDG_DATA_HOME?.trim() ? join(env.XDG_DATA_HOME, "dbfox") : join(homedir(), ".local", "share", "dbfox");
    return join(root, "dlcs", "packages");
  }
  return null;
}

function parseProjection(value: unknown): RuntimeDlcActivationProjection {
  if (value === null || typeof value !== "object") throw new Error("Invalid DLC activation projection");
  const record = value as Record<string, unknown>;
  if (!hasOnlyKeys(record, ["snapshot_id", "active_dlcs"])
    || typeof record.snapshot_id !== "string" || record.snapshot_id.length === 0 || record.snapshot_id.length > 256
    || !Array.isArray(record.active_dlcs) || record.active_dlcs.length > 256) {
    throw new Error("Invalid DLC activation projection");
  }
  const active_dlcs = record.active_dlcs.map((item) => {
    if (item === null || typeof item !== "object") throw new Error("Invalid active DLC item");
    const entry = item as Record<string, unknown>;
    if (!hasOnlyKeys(entry, ["dlc_id", "package_version", "package_digest", "frontend_entrypoint"])
      || typeof entry.dlc_id !== "string" || entry.dlc_id.length === 0 || entry.dlc_id.length > 256
      || typeof entry.package_version !== "string" || entry.package_version.length === 0 || entry.package_version.length > 128
      || typeof entry.package_digest !== "string"
      || !/^[0-9a-f]{64}$/.test(normalizeDigest(entry.package_digest))
      || !(entry.frontend_entrypoint === null || (typeof entry.frontend_entrypoint === "string"
        && entry.frontend_entrypoint.length <= 2_048))) {
      throw new Error("Invalid active DLC item");
    }
    return {
      dlc_id: entry.dlc_id,
      package_version: entry.package_version,
      package_digest: entry.package_digest,
      frontend_entrypoint: entry.frontend_entrypoint,
    };
  });
  return { snapshot_id: record.snapshot_id, active_dlcs };
}

function hasOnlyKeys(record: Record<string, unknown>, allowed: readonly string[]): boolean {
  return Object.keys(record).every((key) => allowed.includes(key));
}

function normalizeDigest(value: string): string {
  return value.replace(/^sha256[:-]/i, "").toLowerCase();
}

function corsHeaders(cacheControl: string): Headers {
  return new Headers({
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "GET, HEAD, OPTIONS",
    "Access-Control-Allow-Headers": "*",
    "Cache-Control": cacheControl,
    "X-Content-Type-Options": "nosniff",
  });
}

function textResponse(status: number, body: string): Response {
  const headers = corsHeaders("no-cache");
  headers.set("Content-Type", "text/plain; charset=utf-8");
  return new Response(body, { status, headers });
}

function mimeForPath(path: string): string {
  const extension = path.split(".").pop()?.toLowerCase();
  return ({
    js: "text/javascript; charset=utf-8", mjs: "text/javascript; charset=utf-8", css: "text/css; charset=utf-8",
    json: "application/json; charset=utf-8", svg: "image/svg+xml", png: "image/png", jpg: "image/jpeg",
    jpeg: "image/jpeg", webp: "image/webp", gif: "image/gif", ico: "image/x-icon", woff: "font/woff",
    woff2: "font/woff2", ttf: "font/ttf", otf: "font/otf", html: "text/html; charset=utf-8",
  } as Record<string, string>)[extension ?? ""] ?? "application/octet-stream";
}

function isNodeError(error: unknown, code: string): boolean {
  return error instanceof Error && "code" in error && error.code === code;
}
